### PR TITLE
selu activation layer

### PR DESCRIPTION
**Pull request by srk97**
This pull request intends to integrate a new activation layer "Scaled Exponential Linear Units" (https://arxiv.org/abs/1706.02515). 
selu
This layer improves the training of deep Feed Forward Neural Networks significantly by maintaining standard normal distribution in the inputs. The dropout has a different implementation  customized for this activation unit. My next pull request will be the implementation of the modified dropout.
```c++
network<sequential> net;
net << conv(32, 32, 5, 1, 6, padding::same) << selu()
      << max_pool(32, 32, 6, 2) << selu()
      << conv(16, 16, 5, 6, 16, padding::same) << selu()
      << max_pool(16, 16, 16, 2) << selu()
      << fc(8*8*16, 100) << selu(100,(float_t)1.1,1.2)
      << fc(100, 10) << softmax(); 
```

### CODE REVIEW

**Comment by Randl** [comment]
Can you add also some example of its usage?

**Comment by srk97** [comment]
Sure, I still have some errors to fix. I'll do that as soon as I fix these errors

**Comment by srk97** [comment]
@Randl What does "coverage decreased" mean? How can I fix it?

File: tiny_dnn/activations/selu_layer.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+#pragma once
+#include "tiny_dnn/activations/activation_layer.h"
+#include "tiny_dnn/layers/layer.h"
+
+namespace tiny_dnn{
+
+	class selu_layer : public activation_layer{
+	public:
+		
+		selu_layer(const float_t lambda=1.05070,const float_t alpha=1.67326)
```
Comment (edgarriba) [inline tiny_dnn/activations/selu_layer.h]: @srk97 please add documentation. Follow this as a reference:
https://github.com/tiny-dnn/tiny-dnn/blob/master/tiny_dnn/activations/leaky_relu_layer.h

**Comment by srk97** [comment]
@edgarriba I've made the changes. I'll add the paper reference and details to the docs once the dropout has been implemented. 
@Randl Added example usage of the layer

**Comment by srk97** [comment]
@edgarriba @Randl  I assume that the drop in coverage is okay since it's only 0.2%.

**Comment by bhack** [comment]
I think that we could have a little training example when dropout is ready to validate the loss curve

**Comment by bhack** [comment]
I agree with @Randl

**Comment by srk97** [comment]
@Randl @bhack  There seems to be a merge conflict in the `serialization_functions.h` file. What do I need to change?

**Comment by Randl** [comment]
Revise on master

**Comment by srk97** [comment]
@bhack @Randl  I tested the calculations locally. I am not familiar with writing tests.

**Comment by srk97** [comment]
@Randl Are there any changes required or Is it ready to merge?

**Comment by edgarriba** [comment]
@srk97 follow this https://github.com/tiny-dnn/tiny-dnn/blob/7d5ce1307a38da18641ee51dba72ecdaba0da4cf/test/test_convolutional_layer.h#L558

Actually I realized that we don't have explicit tests for the new activation layers. Can you create a new test file for this? You can start from here a PR series to add tests for all activation.

**Comment by bhack** [comment]
@edgarriba Yes i noted this yesterday could be very useful to start to cover activation.

**Comment by srk97** [comment]
@edgarriba @bhack Sure. I'll use the tests for reference as I haven't written any tests before. Merge this after the tests are introduced or before?

**Comment by bhack** [comment]
I think you can write this here then for other activations in another PR/PRs

**Comment by srk97** [comment]
```c++
#include "gtest/gtest.h"
#include "testhelper.h"
#include "tiny_dnn/tiny_dnn.h"


using namespace tiny_dnn::activation;

namespace tiny_dnn{
	TEST(selu, gradient_check) {  
  network<sequential> nn;
  nn << selu((serial_size_t)3,(serial_size_t)3,(serial_size_t)1);

  const auto test_data = generate_gradient_check_data(nn.in_data_size());
  nn.init_weight();
 extern bool value= EXPECT_TRUE(nn.gradient_check<mse>(test_data.first, test_data.second,
                                     epsilon<float_t>(), GRAD_CHECK_ALL));

  
}
}
```

@edgarriba @bhack I basically reused the test for convolutional layer by making the necessary modifications. Though I checked the `fprop` and `bprop` manually, I think I need to feed sample data manually and check the output in a test. Is the test above good enough? Forgive me if I missed something.

**Comment by edgarriba** [comment]
@srk97 yeah, go ahead with this

File: tiny_dnn/activations/selu_layer.h [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    return std::make_pair(float_t(0.1), float_t(0.9));
+  }
+
+  float_t lambda_;
+  float_t alpha_;
+
+  friend struct serialization_buddy;
+};
+
+}  // namespace tiny_dnn
```
Comment (beru) [inline tiny_dnn/activations/selu_layer.h]: Please add a new line at the end of the file.

File: test/test_selu_layer.h [hunk 3]
```diff
@@ -0,0 +1,18 @@
+//#pragma once
```
Comment (edgarriba) [inline test/test_selu_layer.h]: uncomment `pragma once`

File: test/test_selu_layer.h [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+#include "gtest/gtest.h"
+#include "testhelper.h"
+#include "tiny_dnn/tiny_dnn.h"
+
+using namespace tiny_dnn::activation;
+
+namespace tiny_dnn {
+TEST(selu, gradient_check) {
+  network<sequential> nn;
+  nn << selu((serial_size_t)3, (serial_size_t)3, (serial_size_t)1);
```
Comment (edgarriba) [inline test/test_selu_layer.h]: replace by `nn << selu(3, 3, 1);`

File: tiny_dnn/util/serialization_functions.h [hunk 5]
```diff
@@ -773,10 +799,11 @@ struct serialization_buddy {
 #endif
 };
 
+
 template <class Archive, typename T>
 typename std::enable_if<std::is_base_of<tiny_dnn::layer, T>::value>::type
 serialize(Archive &ar, T &layer) {
-  serialization_buddy::serialize(ar, layer);
+    serialization_buddy::serialize(ar, layer);
```
Comment (beru) [inline tiny_dnn/util/serialization_functions.h]: This indentation change isn't required.

File: tiny_dnn/util/serialization_functions.h [hunk 6]
```diff
@@ -773,10 +799,11 @@ struct serialization_buddy {
 #endif
 };
 
+
```
Comment (beru) [inline tiny_dnn/util/serialization_functions.h]: This new line isn't probablly needed.

File: tiny_dnn/util/serialization_functions.h [hunk 7]
```diff
@@ -776,7 +802,7 @@ struct serialization_buddy {
 template <class Archive, typename T>
 typename std::enable_if<std::is_base_of<tiny_dnn::layer, T>::value>::type
 serialize(Archive &ar, T &layer) {
-  serialization_buddy::serialize(ar, layer);
+   serialization_buddy::serialize(ar, layer);
```
Comment (beru) [inline tiny_dnn/util/serialization_functions.h]: Please use 2x spaces for indentations.

File: tiny_dnn/util/serialization_functions.h [hunk 8]
```diff
@@ -752,6 +769,15 @@ struct serialization_buddy {
   }
 
   template <class Archive>
+  static inline void serialize(Archive &ar, tiny_dnn::selu_layer &layer) {
+    layer.serialize_prolog(ar);
+    ar(cereal::make_nvp("in_size", layer.in_shape()[0]),
+       cereal::make_nvp("lambda", layer.lambda_),
+       cereal::make_nvp("alpha",layer.alpha_)
+       );
+}
```
Comment (beru) [inline tiny_dnn/util/serialization_functions.h]: This `}` should be indented. Have you used `clang-format` command?
It should automatically indent the code.

File: tiny_dnn/activations/selu_layer.h (hunk 2 above)
Comment (srk97) [inline tiny_dnn/activations/selu_layer.h]: I am sorry. I keep forgetting to clang format the code. Fixed it

**Comment by srk97** [comment]
@edgarriba @beru  Fixed formatting issues. Also, one PR per activation layer test or since they're all in one file, can I make that a single PR?

**Comment by srk97** [comment]
@edgarriba merge this?
